lib/obj.c

```diff
@@ -1441,3 +1442,93 @@ grn_obj_to_script_syntax(grn_ctx *ctx, grn_obj *obj, grn_obj *buffer)
   }
   GRN_API_RETURN(ctx->rc);
 }
+
+static void grn_obj_warm_dispatch(grn_ctx *ctx, grn_obj *obj);
+
+static void
+grn_db_warm(grn_ctx *ctx, grn_obj *obj)
+{
+  grn_db *db = (grn_db *)obj;
+  grn_obj_warm_dispatch(ctx, db->keys);
+  grn_ja_warm(ctx, db->specs);
+  grn_hash_warm(ctx, db->config);
+  grn_options_warm(ctx, db->options);
+
+  GRN_TABLE_EACH_BEGIN_FLAGS(ctx, obj, cursor, id, GRN_CURSOR_BY_ID) {
+    if (id < GRN_N_RESERVED_TYPES) {
+      continue;
+    }
+
+    grn_obj *object = grn_ctx_at(ctx, id);
+    if (!object) {
+      continue;
+    }
+    if (grn_obj_is_table(ctx, object)) {
+      grn_obj_warm_dispatch(ctx, object);
+    }
+    grn_obj_unref(ctx, object);
+  } GRN_TABLE_EACH_END(ctx, cursor);
+}
+
+static void
+grn_table_warm(grn_ctx *ctx, grn_obj *obj)
+{
+  grn_hash *columns = grn_table_all_columns(ctx, obj);
+  if (!columns) {
+    return;
+  }
+
+  GRN_HASH_EACH_BEGIN(ctx, columns, cursor, id) {
+    void *key;
+    grn_hash_cursor_get_key(ctx, cursor, &key);
+    grn_id column_id = *((grn_id *)key);
+    grn_obj *column = grn_ctx_at(ctx, column_id);
+    grn_obj_warm_dispatch(ctx, column);
+    grn_obj_unref(ctx, column);
+  } GRN_HASH_EACH_END(ctx, cursor);
+}
+
+static void
+grn_obj_warm_dispatch(grn_ctx *ctx, grn_obj *obj)
+{
+  switch (obj->header.type) {
+  case GRN_DB :
+    grn_db_warm(ctx, obj);
+    break;
+  case GRN_TABLE_HASH_KEY :
+    grn_hash_warm(ctx, (grn_hash *)obj);
```

Comment:
Don't we need check a return value?
Do we need execute `grn_table_warm` even if `grn_hash_warm` fails?

Reply:
OK. I'll add more return value checks.